### PR TITLE
Align Guice and ASM dependencies

### PR DESCRIPTION
**Pull request by basil**
This PR contains two commits, the second of which depends on the first:

- The first commit adds ASM to the core BOM. This is splitting out a portion of #5979 that is ready to push now, as the rest of #5979 is on-hold. Pushing this portion now reduces risk. It is also necessary for the commit that follows it in order to avoid Enforcer warnings.
- The second commit aligns our Guice dependencies by pulling in `guice-bom`. While we currently only ship `guice` (so this has no effect on the WAR), plugins might depend on e.g. `guice-assistedinject` (as does `artifact-manager-s3-plugin`) and get an arbitrarily older version that is incompatible with the version of Guice we ship in core. By including the full `guice-bom` in the core BOM, we ensure that plugins (which pull in a core BOM corresponding to their baseline) pull in a version of e.g. `guice-assistedinject` that aligns with the version of Guice in their core baseline. Note that this is relying on Guice maintaining API compatibility going forward, which is not necessarily a given. We might need to go even further and bundle `guice-assistedinject` into core, but that is a more drastic step. This is a more conservative first step, and if it's good enough then there's no need to go further and bundle more JARs in core (something we'd rather avoid to decrease maintenance burden).

### Proposed changelog entries

Developer: Plugins now have their Guice dependencies aligned with core.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by timja — APPROVED** [review]

**Review by res0nance — APPROVED** [review]

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

File: core/pom.xml
Comment (jtnord) [inline core/pom.xml:109]: @basil what was the reasoning to add these dependencies here now?
The versions should be aligned from the bom change in the same PR, so should not need to be directly added to the dependencies to get the right version?  I doubt we are directly using all of them - so would only expect a dependency on something we directly used in the code rather than anything that was also transitive?

Comment (timja) [reply core/pom.xml:109]: until it's fully split it's currently part of the API really.

plugins like token macro depend on ASM. and I think scm-api does as well.

Comment (jtnord) [reply core/pom.xml:109]: I am not following why we add these specifically yet are not adding the guice dependencies that some plugins also need (and blow up spectacurly if they get differing version in a plugin).

in other words there it appears to me we have a lack of consistency on what dependencies we add and what we do.

or to put it another way: why is ok for artifact-manager-s3 to use the bom to get a bundle a compatable version of `guice-inject-assist` but not for `token-macro` to use to get the right version of the `asm` library it requires?

Comment (timja) [reply core/pom.xml:109]: I think it's as simple as `guice` publishes a bom and `asm` doesn't

Comment (jglick) [reply core/pom.xml:109]: > I am not following why we add these specifically

Same.

Comment (timja) [reply core/pom.xml:109]: I think it's a matter of carefully removing dependencies, there's a PR to remove JNR: https://github.com/jenkinsci/jenkins/pull/5979

ASM comes via JNR right now, but removing ASM is another task which requires more changes.
So take steps at a time so that we can one day safely remove it, but currently plugins are getting asm libs from core so this means they don't get removed when JNR goes

Comment (basil) [reply core/pom.xml:109]: As Tim mentioned above, the reason is to keep the contents of the WAR identical with regard to the ASM JARs before and after the removal of the JNR JARs in #5979. Removing ASM from the WAR is trickier than removing JNR, so let's go one step at a time.

Comment (basil) [reply core/pom.xml:109]: > plugins like token macro depend on ASM. and I think scm-api does as well.

Core also depends on ASM via [Stapler](https://github.com/jenkinsci/stapler/blob/9a7b4b5916afe4ff0494698cf9ead253bdefd02c/core/src/main/java/org/kohsuke/stapler/ClassDescriptor.java#L28-L33) and [`access-modifier`](https://github.com/jenkinsci/lib-access-modifier/blob/b68c0122e7cfc4772fb2f8ab485b2157fdfa53f8/access-modifier-annotation/src/main/java/org/kohsuke/accmod/restrictions/ProtectedExternally.java#L64-L78), the latter of which [has declared its ASM dependency as `provided`](https://github.com/jenkinsci/lib-access-modifier/blob/b68c0122e7cfc4772fb2f8ab485b2157fdfa53f8/access-modifier-annotation/pom.xml#L19-L23) for five years.

**Pull request merged**
